presto-main/src/main/java/com/facebook/presto/metadata/QueryAwareFunctionAndTypeManager.java

```diff
@@ -0,0 +1,83 @@
+/*
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package com.facebook.presto.metadata;
+
+import com.facebook.presto.common.QualifiedObjectName;
+import com.facebook.presto.common.type.DistinctType;
+import com.facebook.presto.common.type.DistinctTypeInfo;
+import com.facebook.presto.common.type.Type;
+import com.facebook.presto.common.type.TypeSignature;
+import com.facebook.presto.common.type.TypeWithName;
+import com.facebook.presto.common.type.UserDefinedType;
+
+import javax.annotation.concurrent.ThreadSafe;
+
+import java.util.concurrent.ConcurrentHashMap;
+
+@ThreadSafe
+public class QueryAwareFunctionAndTypeManager
```

Comment:
Do we really need this class? Can we bake it into `FunctionAndTypeManager` directly?

Reply:
If we do that, we will have to maintain Map<QueryId, Map<QualifiedObjectName, UserDefinedType>>. When we call getType(), we dont have any context for the query to get the query id.

Reply:
If we do that, we will need Map<queryId, Map<name, type>>, but we dont have any queryid context when calling getType().

Reply:
Discussed offline; let's leverage FunctionAndTypeManager only